File: msword-library/src/main/java/com/deb452/msword_library/EditorComponent.java
package com.deb452.msword_library;

import android.view.View;

import com.deb452.msword_library.components_classes.ComponentsWrapper;
import com.deb452.msword_library.models.EditorContent;
import com.deb452.msword_library.models.EditorType;
import com.deb452.msword_library.models.Node;

import org.jsoup.nodes.Element;

import java.util.ArrayList;

public abstract class EditorComponent {
    private final EditorCore editorCore;
    protected ComponentsWrapper componentsWrapper;
    public abstract Node getContent(View view);
    public abstract String getContentAsHTML(Node node, EditorContent content);
    public abstract void renderEditorFromState(Node node, EditorContent content);
    public abstract Node buildNodeFromHTML(Element element);
    public abstract void init(ComponentsWrapper componentsWrapper);

    public EditorComponent(EditorCore editorCore){
        this.editorCore = editorCore;
    }

    protected Node getNodeInstance(View view){
        Node node = new Node();
        EditorType type = editorCore.getControlType(view);
        node.type = type;
        node.content = new ArrayList<>();
        return node;
    }
    protected Node getNodeInstance(EditorType type){
        Node node = new Node();
        node.type = type;
        node.content = new ArrayList<>();
        return node;
    }

}

File: msword-library/src/main/java/com/deb452/msword_library/EditorComponent.kt
package com.deb452.msword_library

import android.view.View
import com.deb452.msword_library.EditorCore
import com.deb452.msword_library.components_classes.ComponentsWrapper
import com.deb452.msword_library.models.EditorContent
import com.deb452.msword_library.models.EditorType
import com.deb452.msword_library.models.Node
import org.jsoup.nodes.Element
import java.util.ArrayList

abstract class EditorComponent(private val editorCore: EditorCore) {
    @JvmField
    protected var componentsWrapper: ComponentsWrapper? = null
    abstract fun getContent(view: View?): Node?
    abstract fun getContentAsHTML(
        node: Node?,
        content: EditorContent?
    ): String?

    abstract fun renderEditorFromState(
        node: Node?,
        content: EditorContent?
    )

    abstract fun buildNodeFromHTML(element: Element?): Node?
    abstract fun init(componentsWrapper: ComponentsWrapper?)
    protected fun getNodeInstance(view: View?): Node {
        val node = Node()
        val type = editorCore.getControlType(view)
        node.type = type
        node.content = ArrayList()
        return node
    }

    protected fun getNodeInstance(type: EditorType?): Node {
        val node = Node()
        node.type = type
        node.content = ArrayList()
        return node
    }

}